provider/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptClassPathProvider.kt

```diff
@@ -173,3 +189,53 @@ private
 fun isKotlinJar(name: String): Boolean =
     name.startsWith("kotlin-stdlib-")
         || name.startsWith("kotlin-reflect-")
+
+
+private
+class ClassLoaderClassPathCache {
+
+    private
+    val cachedClassPaths = hashMapOf<ClassLoader, ClassPath>()
+
+    fun of(classLoader: ClassLoader): ClassPath =
+        cachedClassPaths.getOrPut(classLoader) {
+            classPathOf(classLoader)
+        }
+
+    private
+    fun classPathOf(classLoader: ClassLoader): ClassPath {
+        val classPathFiles = mutableListOf<File>()
+
+        object : ClassLoaderVisitor() {
+            override fun visitClassPath(classPath: Array<URL>) {
+                classPath.forEach { url ->
+                    if (url.protocol == "file") {
+                        classPathFiles.add(fileFrom(url))
+                    }
```

Comment:
Are there any other kinds of protocols used by the class-loader besides "file"?

Reply:
> Are there any other kinds of protocols used by the class-loader besides "file"?

This is not related to this change which is a port from [ClasspathUtil.getClasspath](https://github.com/gradle/gradle/blob/96f28e5c54616f1af0fe38f943c302a0fd01f55d/subprojects/base-services/src/main/java/org/gradle/internal/classloader/ClasspathUtil.java#L57) with caching added thus the resulting `ClassPath` will be exactly the same as before.

Reply:
Understandable, I just wasn't aware that you could have the class loader work from anywhere but a file.